File: backend/users/roles.py
from typing import Optional

CANONICAL_EXPERT_SLUG = "expert"
CANONICAL_VISITOR_SLUG = "visitor"

ROLE_SLUG_ALIASES = {
    "doctor": CANONICAL_EXPERT_SLUG,
    "expert": CANONICAL_EXPERT_SLUG,
    "patient": CANONICAL_VISITOR_SLUG,
    "visitor": CANONICAL_VISITOR_SLUG,
}


def normalize_role_slug(slug: Optional[str]) -> Optional[str]:
    if not slug:
        return slug
    return ROLE_SLUG_ALIASES.get(slug, slug)


def is_staff_or_admin_user(user) -> bool:
    return bool(
        getattr(user, "is_authenticated", True)
        and (getattr(user, "is_staff", False) or getattr(user, "is_superuser", False))
    )


def is_expert(user) -> bool:
    if is_staff_or_admin_user(user):
        return True
    role = getattr(user, "role", None)
    slug = getattr(role, "slug", None)
    return normalize_role_slug(slug) == CANONICAL_EXPERT_SLUG


def is_visitor(user) -> bool:
    if is_staff_or_admin_user(user):
        return True
    role = getattr(user, "role", None)
    slug = getattr(role, "slug", None)
    return normalize_role_slug(slug) == CANONICAL_VISITOR_SLUG


def has_visitor_features(user) -> bool:
    """
    Visitor features are available to pure visitors and upgraded experts.
    """
    return is_visitor(user) or is_expert(user)

File: frontend/lib/roles.ts
export const CANONICAL_EXPERT_ROLE = "expert";
export const CANONICAL_VISITOR_ROLE = "visitor";

export const ROLE_SLUG_ALIASES: Record<string, string> = {
  doctor: CANONICAL_EXPERT_ROLE,
  expert: CANONICAL_EXPERT_ROLE,
  patient: CANONICAL_VISITOR_ROLE,
  visitor: CANONICAL_VISITOR_ROLE,
};

export function normalizeRoleSlug(roleSlug?: string | null): string | undefined {
  if (!roleSlug) return undefined;
  return ROLE_SLUG_ALIASES[roleSlug] || roleSlug;
}

export function isExpertRoleSlug(roleSlug?: string | null): boolean {
  return normalizeRoleSlug(roleSlug) === CANONICAL_EXPERT_ROLE;
}

export function isVisitorRoleSlug(roleSlug?: string | null): boolean {
  return normalizeRoleSlug(roleSlug) === CANONICAL_VISITOR_ROLE;
}

export function isStaffOrAdminUser(user?: { is_staff?: boolean | null; is_superuser?: boolean | null } | null): boolean {
  return Boolean(user?.is_staff || user?.is_superuser);
}

export function hasExpertFeatures(user?: { role_slug?: string | null; role?: string | null; is_staff?: boolean | null; is_superuser?: boolean | null } | null): boolean {
  return isStaffOrAdminUser(user) || isExpertRoleSlug(user?.role_slug) || isExpertRoleSlug(user?.role);
}

export function hasVisitorFeatures(user?: { role_slug?: string | null; role?: string | null; is_staff?: boolean | null; is_superuser?: boolean | null } | null): boolean {
  return isStaffOrAdminUser(user) || isVisitorRoleSlug(user?.role_slug) || isVisitorRoleSlug(user?.role) || hasExpertFeatures(user);
}
